File: ios/Sources/AssetCachePlugin/AssetCache.swift
import CryptoKit
import Foundation

private enum AssetCacheError: LocalizedError {
    case missingIdentifier
    case invalidUrl
    case invalidResponse
    case httpStatus(Int)

    var errorDescription: String? {
        switch self {
        case .missingIdentifier:
            return "key or url is required"
        case .invalidUrl:
            return "Invalid URL"
        case .invalidResponse:
            return "Invalid HTTP response"
        case .httpStatus(let status):
            return "Failed to fetch asset: HTTP \(status)"
        }
    }
}

private struct StoredAsset: Codable {
    var key: String
    var url: String
    var fileName: String
    var mimeType: String?
    var etag: String?
    var lastModified: String?
    var size: Int64
    var updatedAt: Int64
    var checkedAt: Int64
}

@objc public class AssetCache: NSObject {
    private let rootName = "CapgoAssetCache"

    func get(url urlString: String, key rawKey: String?, headers: [String: String], revalidate: [String: Any]) async throws -> [String: Any] {
        guard let url = URL(string: urlString), let scheme = url.scheme, ["http", "https"].contains(scheme.lowercased()) else {
            throw AssetCacheError.invalidUrl
        }

        try ensureDirectories()

        let key = try resolveKey(rawKey, urlString)
        let strategy = revalidate["strategy"] as? String ?? "never"
        let metadata = try readMetadata(key)

        if let metadata, fileExists(metadata), isFresh(metadata, strategy: strategy, maxAgeSeconds: revalidate["maxAgeSeconds"] as? Int) {
            return result(metadata, fromCache: true, status: "hit")
        }

        var request = URLRequest(url: url)
        for (header, value) in headers {
            request.setValue(value, forHTTPHeaderField: header)
        }
        if let metadata {
            applyValidators(metadata, to: &request, strategy: strategy)
        }

        let (temporaryURL, response) = try await URLSession.shared.download(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw AssetCacheError.invalidResponse
        }

        let now = timestamp()
        if httpResponse.statusCode == 304, var metadata {
            metadata.checkedAt = now
            try writeMetadata(metadata)
            return result(metadata, fromCache: true, status: "notModified")
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            throw AssetCacheError.httpStatus(httpResponse.statusCode)
        }

        let fileName = metadata?.fileName ?? fileName(for: key, url: url)
        let destination = filesDirectory().appendingPathComponent(fileName, isDirectory: false)
        if FileManager.default.fileExists(atPath: destination.path) {
            try FileManager.default.removeItem(at: destination)
        }
        try FileManager.default.moveItem(at: temporaryURL, to: destination)

        let size = try fileSize(destination)
        let next = StoredAsset(
            key: key,
            url: urlString,
            fileName: fileName,
            mimeType: normalizeMimeType(httpResponse.value(forHTTPHeaderField: "Content-Type")),
            etag: httpResponse.value(forHTTPHeaderField: "ETag"),
            lastModified: httpResponse.value(forHTTPHeaderField: "Last-Modified"),
            size: size,
            updatedAt: now,
            checkedAt: now
        )
        try writeMetadata(next)
        return result(next, fromCache: false, status: "downloaded")
    }

    func remove(key rawKey: String?, url urlString: String?) throws -> [String: Any] {
        let key = try resolveKey(rawKey, urlString)
        guard let metadata = try readMetadata(key) else {
            return ["removed": false]
        }

        var removed = false
        let fileURL = fileURL(for: metadata)
        if FileManager.default.fileExists(atPath: fileURL.path) {
            try FileManager.default.removeItem(at: fileURL)
            removed = true
        }
        let metadataURL = metadataURL(for: key)
        if FileManager.default.fileExists(atPath: metadataURL.path) {
            try FileManager.default.removeItem(at: metadataURL)
            removed = true
        }
        return ["removed": removed]
    }

    func clear() throws -> [String: Any] {
        let removed = try storedAssets().count
        let root = rootDirectory()
        if FileManager.default.fileExists(atPath: root.path) {
            try FileManager.default.removeItem(at: root)
        }
        try ensureDirectories()
        return ["removed": removed]
    }

    func list() throws -> [String: Any] {
        let assets = try storedAssets().map { result($0, fromCache: true, status: "hit") }
        return ["assets": assets]
    }

    func getCacheSize() throws -> [String: Any] {
        let size = try storedAssets().reduce(Int64(0)) { total, asset in
            total + (try fileSize(fileURL(for: asset)))
        }
        return ["size": size]
    }

    @objc public func getPluginVersion() -> String {
        return "native"
    }

    private func resolveKey(_ key: String?, _ url: String?) throws -> String {
        if let key, !key.isEmpty {
            return sanitize(key)
        }
        guard let url, !url.isEmpty else {
            throw AssetCacheError.missingIdentifier
        }
        return sha256(url)
    }

    private func isFresh(_ asset: StoredAsset, strategy: String, maxAgeSeconds: Int?) -> Bool {
        switch strategy {
        case "never":
            return true
        case "ttl":
            let ttl = Int64(max(0, maxAgeSeconds ?? 0)) * 1000
            return ttl > 0 && timestamp() - asset.updatedAt < ttl
        default:
            return false
        }
    }

    private func applyValidators(_ asset: StoredAsset, to request: inout URLRequest, strategy: String) {
        if (strategy == "always" || strategy == "etag") && request.value(forHTTPHeaderField: "If-None-Match") == nil {
            request.setValue(asset.etag, forHTTPHeaderField: "If-None-Match")
        }
        if (strategy == "always" || strategy == "last-modified") && request.value(forHTTPHeaderField: "If-Modified-Since") == nil {
            request.setValue(asset.lastModified, forHTTPHeaderField: "If-Modified-Since")
        }
        if strategy == "ttl" {
            if request.value(forHTTPHeaderField: "If-None-Match") == nil {
                request.setValue(asset.etag, forHTTPHeaderField: "If-None-Match")
            }
            if request.value(forHTTPHeaderField: "If-Modified-Since") == nil {
                request.setValue(asset.lastModified, forHTTPHeaderField: "If-Modified-Since")
            }
        }
    }

    private func result(_ asset: StoredAsset, fromCache: Bool, status: String) -> [String: Any] {
        let file = fileURL(for: asset)
        var data: [String: Any] = [
            "key": asset.key,
            "url": asset.url,
            "path": file.path,
            "uri": file.absoluteString,
            "size": asset.size,
            "updatedAt": asset.updatedAt,
            "checkedAt": asset.checkedAt,
            "fromCache": fromCache,
            "status": status
        ]
        data["mimeType"] = asset.mimeType
        data["etag"] = asset.etag
        data["lastModified"] = asset.lastModified
        return data
    }

    private func storedAssets() throws -> [StoredAsset] {
        let directory = metadataDirectory()
        guard FileManager.default.fileExists(atPath: directory.path) else {
            return []
        }

        let files = try FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)
        return files.compactMap { file in
            guard file.pathExtension == "json",
                  let data = try? Data(contentsOf: file),
                  let asset = try? JSONDecoder().decode(StoredAsset.self, from: data),
                  fileExists(asset) else {
                return nil
            }
            return asset
        }
    }

    private func readMetadata(_ key: String) throws -> StoredAsset? {
        let url = metadataURL(for: key)
        guard FileManager.default.fileExists(atPath: url.path) else {
            return nil
        }
        let data = try Data(contentsOf: url)
        return try JSONDecoder().decode(StoredAsset.self, from: data)
    }

    private func writeMetadata(_ asset: StoredAsset) throws {
        let data = try JSONEncoder().encode(asset)
        try data.write(to: metadataURL(for: asset.key), options: .atomic)
    }

    private func fileExists(_ asset: StoredAsset) -> Bool {
        return FileManager.default.fileExists(atPath: fileURL(for: asset).path)
    }

    private func fileURL(for asset: StoredAsset) -> URL {
        return filesDirectory().appendingPathComponent(asset.fileName, isDirectory: false)
    }

    private func metadataURL(for key: String) -> URL {
        return metadataDirectory().appendingPathComponent("\(key).json", isDirectory: false)
    }

    private func ensureDirectories() throws {
        let directories = [rootDirectory(), filesDirectory(), metadataDirectory()]
        for directory in directories where !FileManager.default.fileExists(atPath: directory.path) {
            try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        }
        var root = rootDirectory()
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        try root.setResourceValues(values)
    }

    private func rootDirectory() -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        return base.appendingPathComponent(rootName, isDirectory: true)
    }

    private func filesDirectory() -> URL {
        return rootDirectory().appendingPathComponent("files", isDirectory: true)
    }

    private func metadataDirectory() -> URL {
        return rootDirectory().appendingPathComponent("metadata", isDirectory: true)
    }

    private func fileName(for key: String, url: URL) -> String {
        if key.contains(".") {
            return key
        }
        let pathExtension = url.pathExtension
        return pathExtension.isEmpty ? key : "\(key).\(pathExtension)"
    }

    private func fileSize(_ url: URL) throws -> Int64 {
        let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
        return attributes[.size] as? Int64 ?? 0
    }

    private func normalizeMimeType(_ value: String?) -> String? {
        return value?.split(separator: ";").first.map { String($0) }
    }

    private func sanitize(_ value: String) -> String {
        let allowed = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._-")
        return String(value.unicodeScalars.map { allowed.contains($0) ? Character($0) : "_" })
    }

    private func sha256(_ value: String) -> String {
        let data = Data(value.utf8)
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    private func timestamp() -> Int64 {
        return Int64(Date().timeIntervalSince1970 * 1000)
    }
}

File: android/src/main/java/app/capgo/asset_cache/AssetCache.java
package app.capgo.asset_cache;

import android.content.Context;
import android.net.Uri;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import java.io.BufferedInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Iterator;
import java.util.Locale;
import org.json.JSONObject;

public class AssetCache {

    private static final String ROOT_DIR = "capgo_asset_cache";
    private static final String FILES_DIR = "files";
    private static final String METADATA_DIR = "metadata";
    private final Context context;

    public AssetCache(Context context) {
        this.context = context;
    }

    public JSObject get(String url, String rawKey, JSObject headers, JSObject revalidate) throws Exception {
        if (url == null || url.isEmpty()) {
            throw new IllegalArgumentException("url is required");
        }

        URL remoteUrl = new URL(url);
        String protocol = remoteUrl.getProtocol().toLowerCase(Locale.ROOT);
        if (!"http".equals(protocol) && !"https".equals(protocol)) {
            throw new IllegalArgumentException("Invalid URL");
        }

        ensureDirectories();
        String key = resolveKey(rawKey, url);
        String strategy = revalidate.optString("strategy", "never");
        StoredAsset metadata = readMetadata(key);

        if (metadata != null && fileExists(metadata) && isFresh(metadata, strategy, revalidate.optLong("maxAgeSeconds", 0))) {
            return toResult(metadata, true, "hit");
        }

        HttpURLConnection connection = (HttpURLConnection) remoteUrl.openConnection();
        connection.setConnectTimeout(30000);
        connection.setReadTimeout(30000);
        connection.setInstanceFollowRedirects(true);
        applyHeaders(connection, headers);
        if (metadata != null) {
            applyValidators(connection, metadata, strategy);
        }

        try {
            int status = connection.getResponseCode();
            long now = timestamp();
            if (status == HttpURLConnection.HTTP_NOT_MODIFIED && metadata != null) {
                metadata.checkedAt = now;
                writeMetadata(metadata);
                return toResult(metadata, true, "notModified");
            }
            if (status < 200 || status > 299) {
                throw new IOException("Failed to fetch asset: HTTP " + status);
            }

            String fileName = metadata != null ? metadata.fileName : fileName(key, remoteUrl);
            File destination = new File(filesDirectory(), fileName);
            File temp = File.createTempFile("asset-cache-", ".tmp", rootDirectory());
            try (
                InputStream input = new BufferedInputStream(connection.getInputStream());
                FileOutputStream output = new FileOutputStream(temp)
            ) {
                byte[] buffer = new byte[8192];
                int read;
                while ((read = input.read(buffer)) != -1) {
                    output.write(buffer, 0, read);
                }
            }

            if (destination.exists() && !destination.delete()) {
                throw new IOException("Unable to replace cached asset");
            }
            if (!temp.renameTo(destination)) {
                copyFile(temp, destination);
                if (!temp.delete()) {
                    temp.deleteOnExit();
                }
            }

            StoredAsset next = new StoredAsset();
            next.key = key;
            next.url = url;
            next.fileName = fileName;
            next.mimeType = normalizeMimeType(connection.getHeaderField("Content-Type"));
            next.etag = connection.getHeaderField("ETag");
            next.lastModified = connection.getHeaderField("Last-Modified");
            next.size = destination.length();
            next.updatedAt = now;
            next.checkedAt = now;
            writeMetadata(next);
            return toResult(next, false, "downloaded");
        } finally {
            connection.disconnect();
        }
    }

    public JSObject remove(String rawKey, String url) throws Exception {
        String key = resolveKey(rawKey, url);
        StoredAsset metadata = readMetadata(key);
        boolean removed = false;

        if (metadata != null) {
            File file = fileFor(metadata);
            if (file.exists()) {
                removed = file.delete();
            }
        }

        File metadataFile = metadataFile(key);
        if (metadataFile.exists()) {
            removed = metadataFile.delete() || removed;
        }

        return new JSObject().put("removed", removed);
    }

    public JSObject clear() throws Exception {
        int removed = storedAssets().length();
        deleteRecursively(rootDirectory());
        ensureDirectories();
        return new JSObject().put("removed", removed);
    }

    public JSObject list() throws Exception {
        return new JSObject().put("assets", storedAssets());
    }

    public JSObject getCacheSize() throws Exception {
        long total = 0;
        File[] files = filesDirectory().listFiles();
        if (files != null) {
            for (File file : files) {
                if (file.isFile()) {
                    total += file.length();
                }
            }
        }
        return new JSObject().put("size", total);
    }

    public String getPluginVersion() {
        return "native";
    }

    private JSArray storedAssets() throws Exception {
        JSArray assets = new JSArray();
        File[] files = metadataDirectory().listFiles((dir, name) -> name.endsWith(".json"));
        if (files == null) {
            return assets;
        }

        for (File file : files) {
            StoredAsset asset = readMetadataFile(file);
            if (asset != null && fileExists(asset)) {
                assets.put(toResult(asset, true, "hit"));
            }
        }
        return assets;
    }

    private boolean isFresh(StoredAsset asset, String strategy, long maxAgeSeconds) {
        if ("never".equals(strategy)) {
            return true;
        }
        if ("ttl".equals(strategy)) {
            long ttl = Math.max(0, maxAgeSeconds) * 1000L;
            return ttl > 0 && timestamp() - asset.updatedAt < ttl;
        }
        return false;
    }

    private void applyHeaders(HttpURLConnection connection, JSObject headers) {
        for (Iterator<String> iterator = headers.keys(); iterator.hasNext(); ) {
            String header = iterator.next();
            String value = headers.optString(header, null);
            if (value != null && !value.isEmpty()) {
                connection.setRequestProperty(header, value);
            }
        }
    }

    private void applyValidators(HttpURLConnection connection, StoredAsset asset, String strategy) {
        if (("always".equals(strategy) || "etag".equals(strategy) || "ttl".equals(strategy)) && asset.etag != null) {
            connection.setRequestProperty("If-None-Match", asset.etag);
        }
        if (("always".equals(strategy) || "last-modified".equals(strategy) || "ttl".equals(strategy)) && asset.lastModified != null) {
            connection.setRequestProperty("If-Modified-Since", asset.lastModified);
        }
    }

    private JSObject toResult(StoredAsset asset, boolean fromCache, String status) {
        JSObject result = new JSObject();
        File file = fileFor(asset);
        result.put("key", asset.key);
        result.put("url", asset.url);
        result.put("path", file.getAbsolutePath());
        result.put("uri", Uri.fromFile(file).toString());
        result.put("mimeType", asset.mimeType);
        result.put("etag", asset.etag);
        result.put("lastModified", asset.lastModified);
        result.put("size", asset.size);
        result.put("updatedAt", asset.updatedAt);
        result.put("checkedAt", asset.checkedAt);
        result.put("fromCache", fromCache);
        result.put("status", status);
        return result;
    }

    private StoredAsset readMetadata(String key) throws Exception {
        File file = metadataFile(key);
        if (!file.exists()) {
            return null;
        }
        return readMetadataFile(file);
    }

    private StoredAsset readMetadataFile(File file) throws Exception {
        String raw;
        try (InputStream input = new FileInputStream(file)) {
            raw = readString(input);
        }
        JSONObject json = new JSONObject(raw);
        StoredAsset asset = new StoredAsset();
        asset.key = json.getString("key");
        asset.url = json.getString("url");
        asset.fileName = json.getString("fileName");
        asset.mimeType = optionalString(json, "mimeType");
        asset.etag = optionalString(json, "etag");
        asset.lastModified = optionalString(json, "lastModified");
        asset.size = json.optLong("size", 0);
        asset.updatedAt = json.optLong("updatedAt", 0);
        asset.checkedAt = json.optLong("checkedAt", 0);
        return asset;
    }

    private String optionalString(JSONObject json, String key) {
        return json.isNull(key) ? null : json.optString(key, null);
    }

    private String readString(InputStream input) throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        byte[] buffer = new byte[8192];
        int read;
        while ((read = input.read(buffer)) != -1) {
            output.write(buffer, 0, read);
        }
        return output.toString(StandardCharsets.UTF_8.name());
    }

    private void writeMetadata(StoredAsset asset) throws Exception {
        JSONObject json = new JSONObject();
        json.put("key", asset.key);
        json.put("url", asset.url);
        json.put("fileName", asset.fileName);
        json.put("mimeType", asset.mimeType);
        json.put("etag", asset.etag);
        json.put("lastModified", asset.lastModified);
        json.put("size", asset.size);
        json.put("updatedAt", asset.updatedAt);
        json.put("checkedAt", asset.checkedAt);
        try (FileOutputStream output = new FileOutputStream(metadataFile(asset.key))) {
            output.write(json.toString().getBytes(StandardCharsets.UTF_8));
        }
    }

    private String resolveKey(String rawKey, String url) throws Exception {
        if (rawKey != null && !rawKey.isEmpty()) {
            return sanitize(rawKey);
        }
        if (url == null || url.isEmpty()) {
            throw new IllegalArgumentException("key or url is required");
        }
        return sha256(url);
    }

    private String sanitize(String value) {
        String sanitized = value.replaceAll("[^a-zA-Z0-9._-]", "_");
        return sanitized.isEmpty() ? "asset" : sanitized;
    }

    private String sha256(String value) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] hash = digest.digest(value.getBytes(StandardCharsets.UTF_8));
        StringBuilder builder = new StringBuilder();
        for (byte item : hash) {
            builder.append(String.format("%02x", item & 0xff));
        }
        return builder.toString();
    }

    private String fileName(String key, URL url) {
        if (key.contains(".")) {
            return key;
        }
        String path = url.getPath();
        int dotIndex = path.lastIndexOf('.');
        if (dotIndex >= 0 && dotIndex < path.length() - 1) {
            return key + path.substring(dotIndex);
        }
        return key;
    }

    private boolean fileExists(StoredAsset asset) {
        return fileFor(asset).exists();
    }

    private File fileFor(StoredAsset asset) {
        return new File(filesDirectory(), asset.fileName);
    }

    private File metadataFile(String key) {
        return new File(metadataDirectory(), key + ".json");
    }

    private File rootDirectory() {
        return new File(context.getFilesDir(), ROOT_DIR);
    }

    private File filesDirectory() {
        return new File(rootDirectory(), FILES_DIR);
    }

    private File metadataDirectory() {
        return new File(rootDirectory(), METADATA_DIR);
    }

    private void ensureDirectories() throws IOException {
        if (!filesDirectory().exists() && !filesDirectory().mkdirs()) {
            throw new IOException("Unable to create asset cache files directory");
        }
        if (!metadataDirectory().exists() && !metadataDirectory().mkdirs()) {
            throw new IOException("Unable to create asset cache metadata directory");
        }
    }

    private void copyFile(File source, File destination) throws IOException {
        try (InputStream input = new FileInputStream(source); FileOutputStream output = new FileOutputStream(destination)) {
            byte[] buffer = new byte[8192];
            int read;
            while ((read = input.read(buffer)) != -1) {
                output.write(buffer, 0, read);
            }
        }
    }

    private void deleteRecursively(File file) {
        if (!file.exists()) {
            return;
        }
        if (file.isDirectory()) {
            File[] children = file.listFiles();
            if (children != null) {
                for (File child : children) {
                    deleteRecursively(child);
                }
            }
        }
        file.delete();
    }

    private String normalizeMimeType(String value) {
        if (value == null) {
            return null;
        }
        int separator = value.indexOf(';');
        return separator == -1 ? value : value.substring(0, separator);
    }

    private long timestamp() {
        return System.currentTimeMillis();
    }

    private static class StoredAsset {

        String key;
        String url;
        String fileName;
        String mimeType;
        String etag;
        String lastModified;
        long size;
        long updatedAt;
        long checkedAt;
    }
}
